### ISSUE TITLE
Ask for help, on CIFAR and CUB200 datasets, re-implementation of DropMax performances worser than baseline softmax

### ISSUE THREAD

**Issue by XueZ-phd** [haebeom-lee/dropmax#7]
Hi Author,
Thanks for your released codes of DropMax on MNIST/LeNet5, it is very convenient for us to understand your paper.

Following your implementations of 'lenet/dropmax' as well as Kaiming He's [CIFAR/resnet(6n+2)](https://arxiv.org/pdf/1512.03385.pdf), I re-implement dropmax on CIFAR10/resnetV2_20, CIFAR100/resnetV2_32 as well as CUB200_2011/resnetV2_50. 

- For **baseline softmax**, after 5 runs, I get _91.30% ± 0.21 (CIFAR10/resnetV2_20) , 68.92% ± 0.25 (CIFAR100/resnetV2_32) and 66.213%±0.224 (CUB200_2011/resnetV2_50)_

- However, for  **DropMax**, after 5 runs, **it doesn't always permance better than baseline softmax, even worse**, specifically: _91.72% ± 0.06 (CIFAR10/resnetV2_20), 67.978% ± 0.435 (CIFAR100/resnetV2_32) and 47.256% ± 0.426 (CUB200_2011/resnetV2_50)_

Why is there such a big difference and I am eager for your help.

During my replication, obeying the network structure in Figure 2 of your paper, I did not modify the code of dropmax in the ph_branch and rh_branch, but simply flowed the features from the penultimate layer of the resnet to ph_branch and rh_branch, and updated the network parameters according to the loss function you designed. 

For dropmax training, I use the same hyperparameters as for softmax training, including the initial learning rate, the number of epochs, etc.

I've taken some of the key codes of my re-implementation on CUB200_2011/resnetV2_50 below:

-  **create dataset**

train_dataset = dataset(args, 'train', image_names, image_splits, image_labels, class_names)
train_dataset.createDataset()
train_label_batch, train_image_batch, train_raw_image_batch, train_name_batch = train_dataset.iterator.get_next()
train_image_batch = tf.reshape(train_image_batch, [args.batch_size, args.image_width_height, args.image_width_height, 3], name='image_batch/train')
train_label_batch = tf.cast(train_label_batch, tf.int64)
iterations_per_epoch = train_dataset.iterations
steps = iterations_per_epoch * args.num_epoch
print("steps per epoch: %d, total training steps: %d" % (iterations_per_epoch, steps))

test_dataset = dataset(args, 'test', image_names, image_splits, image_labels, class_names)
test_dataset.createDataset()
test_label_batch, test_image_batch, test_raw_image_batch, test_name_batch = test_dataset.iterator.get_next()
test_image_batch = tf.reshape(test_image_batch, [-1, args.image_width_height, args.image_width_height, 3], name='image_batch/test')
test_label_batch = tf.cast(test_label_batch, tf.int64)

- **forward**

train_h, train_logits  = GAP_Resnet_v2.get_resnet50_logits(train_image_batch, args.num_cls, False, True, base_scope="resnet_v2_50", FC_scope="resnet_v2_50_Final_Pre") # train_h and train_logits corresponding to the h and o in the DropMax paper Figure 2(b)
test_h, test_logits = GAP_Resnet_v2.get_resnet50_logits(test_image_batch, args.num_cls, True, False, base_scope="resnet_v2_50", FC_scope="resnet_v2_50_Final_Pre") # test_h and test_logits corresponding to the h and o in the DropMax paper Figure 2(c)

- **backward**

global_step = tf.get_variable("global_step", shape=[], dtype=tf.int32, initializer=tf.zeros_initializer, trainable=False)
one_hot_train = tf.one_hot(train_label_batch, args.num_cls, name='train_one_hot')
one_hot_test = tf.one_hot(test_label_batch, args.num_cls, name='test_one_hot')
train_net = drop_tail(one_hot_train, train_h, train_logits, True, args.num_cls, name='resnet_v2_50', reuse=False)
test_net = drop_tail(one_hot_test, test_h, test_logits, False, args.num_cls, name='resnet_v2_50', reuse=True)

train_loss_batch = train_net['cent'] + train_net['kl'] + train_net['aux'] + train_net['neg_ent'] + train_net['wd']

lr = tf.train.piecewise_constant(global_step, [int(0.5*steps), int(0.75*steps)], [1e-2, 1e-3, 1e-4])
optimizer = tf.train.MomentumOptimizer(lr, momentum=0.9)
update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
train_op = slim.learning.create_train_op(train_loss_batch, optimizer, global_step, update_ops)

- **metrics**

train_accuracy_batch = train_net['acc1']
test_accuracy_batch = test_net['acc1']
test_CE_batch = test_net['cent']
`

where, 

- droptail is exact copy of lenet/dropmax
def **drop_tail**(y, h, o, training, num_cls=10, name='dropmax', reuse=None,):
    ph = dense(h, num_cls, name=name + '/ph_branch', reuse=reuse)
    rh = dense(h, num_cls, name=name + '/rh_branch', reuse=reuse)
    qh = tf.stop_gradient(ph) + rh

    p = sigmoid(ph)
    r = sigmoid(rh)
    q = sigmoid(qh)

    _sampling the dropout masks_
    z = genmask(q, y)

    net = {}
    net['cent'] = cross_entropy(((z if training else p) + eps) * exp(o), y)
    net['acc'] = accuracy((p + eps) * exp(o), y)
    net['acc1'] = accuracy1((p + eps) * exp(o), y)
    all_vars = tf.get_collection('variables', scope=name)
    net['weights'] = all_vars
    net['wd'] = weight_decay(1e-4,
                             var_list=[v for v in all_vars if 'branch' not in v.name])

    _dropmax modules_
    net['kl'] = kl_divergence(p, q, y)
    net['aux'] = auxloss(r, y)
    net['neg_ent'] = neg_entropy(p)
    return net

- get_resnet50_logits
def **get_resnet50_logits**(image_batch, num_cls, reuse=False, is_training=False, keep_prob=1.0, base_scope="resnet_v2_50",
                        FC_scope="resnet_v2_50_Fianl_Pre"):
    with slim.arg_scope(resnet_v2.resnet_arg_scope()):
        net, end_points = resnet_v2.resnet_v2_50(image_batch, num_classes=None, is_training=is_training,
                                                 global_pool=True, reuse=reuse, scope=base_scope)
        with tf.variable_scope(FC_scope, reuse=reuse):
            pre = slim.dropout(net, keep_prob=keep_prob, scope='Dropout_1b')
            pre = slim.flatten(pre, scope="flatten_pre")
            shape = pre.get_shape().as_list()
            FC_w_Cx23 = tf.get_variable("FC_Cx23", shape=[shape[-1], num_cls], dtype=tf.float32,
                                        initializer=initializers.xavier_initializer())
            logits = tf.matmul(pre, FC_w_Cx23)  # bs x 23
            return pre, logits
